File: scripts/reimport-rolando.mjs
#!/usr/bin/env node
/**
 * Complete Reimport Script for Rolando Excel Data
 * Run with: node scripts/reimport-rolando.mjs
 */

import XLSX from 'xlsx';
import mysql from 'mysql2/promise';
import * as fs from 'fs';
import * as path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Helper to clean phone numbers
function cleanPhone(phone) {
  if (!phone || phone === '' || phone === 'nan' || phone === 'NaN') return null;
  const phoneStr = String(phone).replace('.0', '').replace(/\.0$/, '').trim();
  if (!phoneStr || phoneStr === 'nan' || phoneStr === 'NaN') return null;
  const cleaned = phoneStr.replace(/[^\d+]/g, '');
  return cleaned || null;
}

// Helper to clean emails
function cleanEmail(email) {
  if (!email || email === '' || email === 'nan' || email === 'NaN') return null;
  const emailStr = String(email).trim().toLowerCase();
  if (!emailStr || emailStr === 'nan' || !emailStr.includes('@')) return null;
  return emailStr;
}

// Parse DATABASE_URL
function parseDbUrl(url) {
  // Remove ssl parameter if present
  const cleanUrl = url.split('?')[0];
  const match = cleanUrl.match(/mysql:\/\/([^:]+):([^@]+)@([^:]+):(\d+)\/(.+)/);
  if (!match) throw new Error('Invalid DATABASE_URL format');
  return {
    user: match[1],
    password: decodeURIComponent(match[2]),
    host: match[3],
    port: parseInt(match[4]),
    database: match[5]
  };
}

async function main() {
  console.log('='.repeat(80));
  console.log('ROLANDO EXCEL COMPLETE REIMPORT');
  console.log('='.repeat(80));
  
  // Read DATABASE_URL from environment or .env file
  let dbUrl = process.env.DATABASE_URL;
  
  if (!dbUrl) {
    // Try to read from .env file
    const envPath = path.join(__dirname, '..', '.env');
    if (fs.existsSync(envPath)) {
      const envContent = fs.readFileSync(envPath, 'utf-8');
      const match = envContent.match(/DATABASE_URL=["']?([^"'\n]+)["']?/);
      if (match) {
        dbUrl = match[1];
      }
    }
  }
  
  if (!dbUrl) {
    console.error('ERROR: DATABASE_URL not found');
    process.exit(1);
  }
  
  // Parse connection details
  const dbConfig = parseDbUrl(dbUrl);
  console.log(`\n🔌 Connecting to database at ${dbConfig.host}:${dbConfig.port}...`);
  
  // Connect to database
  const connection = await mysql.createConnection({
    ...dbConfig,
    ssl: {
      rejectUnauthorized: true
    }
  });
  console.log('   Connected successfully!');
  
  // Load Excel file
  const excelPath = path.join(__dirname, '..', 'dealmachine-properties-2026-01-12-220953_rolando.xlsx');
  console.log(`\n📂 Loading Excel file: ${excelPath}`);
  
  const workbook = XLSX.readFile(excelPath);
  const sheetName = workbook.SheetNames[0];
  const worksheet = workbook.Sheets[sheetName];
  const data = XLSX.utils.sheet_to_json(worksheet);
  console.log(`   Found ${data.length} properties in Excel`);
  
  // Step 1: Get all property IDs with dealMachinePropertyId
  console.log(`\n🔍 Finding existing Rolando properties...`);
  const [existingProps] = await connection.execute(
    'SELECT id, dealMachinePropertyId FROM properties WHERE dealMachinePropertyId IS NOT NULL'
  );
  
  const propMap = new Map();
  existingProps.forEach(p => {
    propMap.set(String(p.dealMachinePropertyId), p.id);
  });
  console.log(`   Found ${propMap.size} existing Rolando properties in database`);
  
  // Step 2: Delete existing contacts, phones, emails for these properties
  if (propMap.size > 0) {
    const propIds = Array.from(propMap.values());
    console.log(`\n🗑️  Cleaning existing contact data...`);
    
    // Get contact IDs for these properties
    const placeholders = propIds.map(() => '?').join(',');
    const [existingContacts] = await connection.execute(
      `SELECT id FROM contacts WHERE propertyId IN (${placeholders})`,
      propIds
    );
    
    const contactIds = existingContacts.map(c => c.id);
    
    if (contactIds.length > 0) {
      const contactPlaceholders = contactIds.map(() => '?').join(',');
      
      // Delete phones
      const [phoneResult] = await connection.execute(
        `DELETE FROM contactPhones WHERE contactId IN (${contactPlaceholders})`,
        contactIds
      );
      console.log(`   Deleted ${phoneResult.affectedRows} phone records`);
      
      // Delete emails
      const [emailResult] = await connection.execute(
        `DELETE FROM contactEmails WHERE contactId IN (${contactPlaceholders})`,
        contactIds
      );
      console.log(`   Deleted ${emailResult.affectedRows} email records`);
      
      // Delete contacts
      const [contactResult] = await connection.execute(
        `DELETE FROM contacts WHERE propertyId IN (${placeholders})`,
        propIds
      );
      console.log(`   Deleted ${contactResult.affectedRows} contact records`);
    }
  }
  
  // Step 3: Import all contacts from Excel
  console.log(`\n📥 Importing contacts from Excel...`);
  
  let totalContacts = 0;
  let totalPhones = 0;
  let totalEmails = 0;
  let propertiesUpdated = 0;
  
  for (let idx = 0; idx < data.length; idx++) {
    const row = data[idx];
    const propertyIdDm = String(row['property_id'] || '');
    
    if (!propMap.has(propertyIdDm)) {
      continue;
    }
    
    const propertyId = propMap.get(propertyIdDm);
    propertiesUpdated++;
    
    // Update property with lat/lng/county/url/flags in dealMachineRawData
    const rawData = {};
    if (row['property_lat']) rawData.lat = Number(row['property_lat']);
    if (row['property_lng']) rawData.lng = Number(row['property_lng']);
    if (row['property_address_county']) rawData.county = String(row['property_address_county']);
    if (row['dealmachine_url']) rawData.dealMachineUrl = String(row['dealmachine_url']);
    
    // Store property flags in rawData
    const propertyFlags = row['property_flags'];
    if (propertyFlags && String(propertyFlags).trim()) {
      const flagsList = String(propertyFlags).split(',').map(f => f.trim()).filter(f => f);
      if (flagsList.length > 0) {
        rawData.propertyFlags = flagsList;
      }
    }
    
    if (Object.keys(rawData).length > 0) {
      await connection.execute(
        'UPDATE properties SET dealMachineRawData = ? WHERE id = ?',
        [JSON.stringify(rawData), propertyId]
      );
    }
    
    // Import ALL 20 contacts
    for (let contactNum = 1; contactNum <= 20; contactNum++) {
      const nameCol = `contact_${contactNum}_name`;
      const flagsCol = `contact_${contactNum}_flags`;
      
      const contactName = row[nameCol];
      if (!contactName || !String(contactName).trim()) {
        continue;
      }
      
      const cleanName = String(contactName).trim();
      const contactFlags = row[flagsCol] ? String(row[flagsCol]) : '';
      
      // Determine relationship from flags
      let relationship = 'Other';
      if (contactFlags.includes('Likely Owner')) {
        relationship = 'Owner';
      } else if (contactFlags.includes('Resident')) {
        relationship = 'Resident';
      } else if (contactFlags.includes('Relative')) {
        relationship = 'Relative';
      }
      
      // Insert contact
      const [contactResult] = await connection.execute(
        'INSERT INTO contacts (propertyId, name, relationship, flags, createdAt, updatedAt) VALUES (?, ?, ?, ?, NOW(), NOW())',
        [propertyId, cleanName, relationship, contactFlags]
      );
      
      const contactId = contactResult.insertId;
      totalContacts++;
      
      // Import all 3 phones
      for (let phoneNum = 1; phoneNum <= 3; phoneNum++) {
        const phoneCol = `contact_${contactNum}_phone${phoneNum}`;
        const phone = cleanPhone(row[phoneCol]);
        if (phone) {
          await connection.execute(
            'INSERT INTO contactPhones (contactId, phoneNumber, createdAt) VALUES (?, ?, NOW())',
            [contactId, phone]
          );
          totalPhones++;
        }
      }
      
      // Import all 3 emails
      for (let emailNum = 1; emailNum <= 3; emailNum++) {
        const emailCol = `contact_${contactNum}_email${emailNum}`;
        const email = cleanEmail(row[emailCol]);
        if (email) {
          await connection.execute(
            'INSERT INTO contactEmails (contactId, email, createdAt) VALUES (?, ?, NOW())',
            [contactId, email]
          );
          totalEmails++;
        }
      }
    }
    
    // Progress update every 50 properties
    if ((idx + 1) % 50 === 0) {
      console.log(`   Processed ${idx + 1}/${data.length} properties...`);
    }
  }
  
  // Step 4: Verify counts
  console.log(`\n✅ IMPORT COMPLETE!`);
  console.log('='.repeat(80));
  console.log(`   Properties updated: ${propertiesUpdated}`);
  console.log(`   Contacts imported: ${totalContacts}`);
  console.log(`   Phones imported: ${totalPhones}`);
  console.log(`   Emails imported: ${totalEmails}`);
  
  // Verify in database
  console.log(`\n🔍 Verifying database counts...`);
  const [[dbContacts]] = await connection.execute('SELECT COUNT(*) as cnt FROM contacts');
  const [[dbPhones]] = await connection.execute('SELECT COUNT(*) as cnt FROM contactPhones');
  const [[dbEmails]] = await connection.execute('SELECT COUNT(*) as cnt FROM contactEmails');
  
  console.log(`   Database contacts: ${dbContacts.cnt}`);
  console.log(`   Database phones: ${dbPhones.cnt}`);
  console.log(`   Database emails: ${dbEmails.cnt}`);
  
  // Close connection
  await connection.end();
  
  console.log(`\n🎉 Reimport completed successfully!`);
}

main().catch(err => {
  console.error('Error:', err);
  process.exit(1);
});

File: reimport_rolando.py
#!/usr/bin/env python3
"""
Complete Reimport Script for Rolando Excel Data
This script will:
1. Delete all existing contacts, phones, emails for Rolando properties
2. Reimport ALL 20 contacts per property with ALL phones and emails
3. Import property_flags as tags
4. Store lat/lng/county/dealMachineUrl properly
"""

import pandas as pd
import mysql.connector
import os
import json
from datetime import datetime

# Database connection
DB_CONFIG = {
    'host': os.environ.get('DB_HOST', 'gateway01.us-east-1.prod.aws.tidbcloud.com'),
    'port': int(os.environ.get('DB_PORT', 4000)),
    'user': os.environ.get('DB_USER'),
    'password': os.environ.get('DB_PASSWORD'),
    'database': os.environ.get('DB_NAME'),
    'ssl_ca': '/etc/ssl/certs/ca-certificates.crt',
    'ssl_verify_cert': True
}

def get_db_connection():
    """Get database connection from environment"""
    # Read from .env file if exists
    env_path = '/home/ubuntu/crm-123drive-v2/.env'
    if os.path.exists(env_path):
        with open(env_path, 'r') as f:
            for line in f:
                if '=' in line and not line.startswith('#'):
                    key, value = line.strip().split('=', 1)
                    os.environ[key] = value.strip('"').strip("'")
    
    # Parse DATABASE_URL
    db_url = os.environ.get('DATABASE_URL', '')
    if db_url:
        # Format: mysql://user:pass@host:port/database
        import re
        match = re.match(r'mysql://([^:]+):([^@]+)@([^:]+):(\d+)/(.+)', db_url)
        if match:
            return mysql.connector.connect(
                host=match.group(3),
                port=int(match.group(4)),
                user=match.group(1),
                password=match.group(2),
                database=match.group(5),
                ssl_ca='/etc/ssl/certs/ca-certificates.crt',
                ssl_verify_cert=True
            )
    
    raise Exception("Could not parse DATABASE_URL")

def clean_phone(phone):
    """Clean phone number - remove .0 suffix and format"""
    if pd.isna(phone) or not phone:
        return None
    phone_str = str(phone).replace('.0', '').strip()
    if not phone_str or phone_str == 'nan':
        return None
    # Remove any non-digit characters except +
    cleaned = ''.join(c for c in phone_str if c.isdigit() or c == '+')
    return cleaned if cleaned else None

def clean_email(email):
    """Clean email address"""
    if pd.isna(email) or not email:
        return None
    email_str = str(email).strip().lower()
    if not email_str or email_str == 'nan' or '@' not in email_str:
        return None
    return email_str

def main():
    print("=" * 80)
    print("ROLANDO EXCEL COMPLETE REIMPORT")
    print("=" * 80)
    
    # Load Excel file
    file_path = '/home/ubuntu/crm-123drive-v2/dealmachine-properties-2026-01-12-220953_rolando.xlsx'
    print(f"\n📂 Loading Excel file: {file_path}")
    df = pd.read_excel(file_path)
    print(f"   Found {len(df)} properties in Excel")
    
    # Connect to database
    print(f"\n🔌 Connecting to database...")
    conn = get_db_connection()
    cursor = conn.cursor(dictionary=True)
    print("   Connected successfully!")
    
    # Step 1: Get all property IDs with dealMachinePropertyId
    print(f"\n🔍 Finding existing Rolando properties...")
    cursor.execute("SELECT id, dealMachinePropertyId FROM properties WHERE dealMachinePropertyId IS NOT NULL")
    existing_props = {str(row['dealMachinePropertyId']): row['id'] for row in cursor.fetchall()}
    print(f"   Found {len(existing_props)} existing Rolando properties in database")
    
    # Step 2: Delete existing contacts, phones, emails for these properties
    if existing_props:
        prop_ids = list(existing_props.values())
        print(f"\n🗑️  Cleaning existing contact data...")
        
        # Get contact IDs for these properties
        format_strings = ','.join(['%s'] * len(prop_ids))
        cursor.execute(f"SELECT id FROM contacts WHERE propertyId IN ({format_strings})", prop_ids)
        contact_ids = [row['id'] for row in cursor.fetchall()]
        
        if contact_ids:
            contact_format = ','.join(['%s'] * len(contact_ids))
            
            # Delete phones
            cursor.execute(f"DELETE FROM contactPhones WHERE contactId IN ({contact_format})", contact_ids)
            phones_deleted = cursor.rowcount
            print(f"   Deleted {phones_deleted} phone records")
            
            # Delete emails
            cursor.execute(f"DELETE FROM contactEmails WHERE contactId IN ({contact_format})", contact_ids)
            emails_deleted = cursor.rowcount
            print(f"   Deleted {emails_deleted} email records")
            
            # Delete contacts
            cursor.execute(f"DELETE FROM contacts WHERE propertyId IN ({format_strings})", prop_ids)
            contacts_deleted = cursor.rowcount
            print(f"   Deleted {contacts_deleted} contact records")
        
        conn.commit()
    
    # Step 3: Import all contacts from Excel
    print(f"\n📥 Importing contacts from Excel...")
    
    total_contacts = 0
    total_phones = 0
    total_emails = 0
    properties_updated = 0
    
    for idx, row in df.iterrows():
        property_id_dm = str(row.get('property_id', ''))
        
        if property_id_dm not in existing_props:
            continue
        
        property_id = existing_props[property_id_dm]
        properties_updated += 1
        
        # Update property with flags as tags
        property_flags = row.get('property_flags', '')
        if pd.notna(property_flags) and property_flags:
            flags_list = [f.strip() for f in str(property_flags).split(',') if f.strip()]
            tags_json = json.dumps(flags_list)
            cursor.execute("UPDATE properties SET tags = %s WHERE id = %s", (tags_json, property_id))
        
        # Update property with lat/lng/county
        lat = row.get('property_lat')
        lng = row.get('property_lng')
        county = row.get('property_address_county')
        dm_url = row.get('dealmachine_url')
        
        raw_data = {}
        if pd.notna(lat):
            raw_data['lat'] = float(lat)
        if pd.notna(lng):
            raw_data['lng'] = float(lng)
        if pd.notna(county):
            raw_data['county'] = str(county)
        if pd.notna(dm_url):
            raw_data['dealMachineUrl'] = str(dm_url)
        
        if raw_data:
            cursor.execute("UPDATE properties SET dealMachineRawData = %s WHERE id = %s", 
                          (json.dumps(raw_data), property_id))
        
        # Import ALL 20 contacts
        for contact_num in range(1, 21):
            name_col = f'contact_{contact_num}_name'
            flags_col = f'contact_{contact_num}_flags'
            
            if name_col not in df.columns:
                continue
            
            contact_name = row.get(name_col)
            if pd.isna(contact_name) or not str(contact_name).strip():
                continue
            
            contact_name = str(contact_name).strip()
            contact_flags = str(row.get(flags_col, '')) if pd.notna(row.get(flags_col)) else ''
            
            # Determine relationship from flags
            relationship = 'Other'
            if 'Likely Owner' in contact_flags:
                relationship = 'Owner'
            elif 'Resident' in contact_flags:
                relationship = 'Resident'
            elif 'Relative' in contact_flags:
                relationship = 'Relative'
            
            # Insert contact
            cursor.execute("""
                INSERT INTO contacts (propertyId, name, relationship, createdAt, updatedAt)
                VALUES (%s, %s, %s, NOW(), NOW())
            """, (property_id, contact_name, relationship))
            contact_id = cursor.lastrowid
            total_contacts += 1
            
            # Import all 3 phones
            for phone_num in range(1, 4):
                phone_col = f'contact_{contact_num}_phone{phone_num}'
                if phone_col in df.columns:
                    phone = clean_phone(row.get(phone_col))
                    if phone:
                        cursor.execute("""
                            INSERT INTO contactPhones (contactId, phoneNumber, createdAt)
                            VALUES (%s, %s, NOW())
                        """, (contact_id, phone))
                        total_phones += 1
            
            # Import all 3 emails
            for email_num in range(1, 4):
                email_col = f'contact_{contact_num}_email{email_num}'
                if email_col in df.columns:
                    email = clean_email(row.get(email_col))
                    if email:
                        cursor.execute("""
                            INSERT INTO contactEmails (contactId, email, createdAt)
                            VALUES (%s, %s, NOW())
                        """, (contact_id, email))
                        total_emails += 1
        
        # Progress update every 50 properties
        if (idx + 1) % 50 == 0:
            print(f"   Processed {idx + 1}/{len(df)} properties...")
            conn.commit()
    
    # Final commit
    conn.commit()
    
    # Step 4: Verify counts
    print(f"\n✅ IMPORT COMPLETE!")
    print("=" * 80)
    print(f"   Properties updated: {properties_updated}")
    print(f"   Contacts imported: {total_contacts}")
    print(f"   Phones imported: {total_phones}")
    print(f"   Emails imported: {total_emails}")
    
    # Verify in database
    print(f"\n🔍 Verifying database counts...")
    cursor.execute("SELECT COUNT(*) as cnt FROM contacts")
    db_contacts = cursor.fetchone()['cnt']
    cursor.execute("SELECT COUNT(*) as cnt FROM contactPhones")
    db_phones = cursor.fetchone()['cnt']
    cursor.execute("SELECT COUNT(*) as cnt FROM contactEmails")
    db_emails = cursor.fetchone()['cnt']
    
    print(f"   Database contacts: {db_contacts}")
    print(f"   Database phones: {db_phones}")
    print(f"   Database emails: {db_emails}")
    
    # Close connection
    cursor.close()
    conn.close()
    
    print(f"\n🎉 Reimport completed successfully!")

if __name__ == '__main__':
    main()
